### PR TITLE
docgen: get rid of ropes

### PR DESCRIPTION
**Pull request by a-mr**
Using strings instead of ropes makes code cleaner. And it runs faster by 5% (measured `nim rst2html manual.rst`).

cc @narimiran

### CODE REVIEW

**Comment by a-mr** [comment]
Checked by `rst2html manual.rst` and `doc os.nim` that there is no difference (besides timestamp) b/w `.html` generated by new and old version.

**Comment by timotheecour** [comment]
CI failure unrelated (fixing it here: https://github.com/nim-lang/Nim/pull/17940)

> And it runs faster by 5% (measured nim rst2html manual.rst)

what about with -d:release?
what about with -d:danger?
(also, make sure it's apples to apples comparison, eg by using `-f` to avoid nimBetterRun kicking in

**Comment by Araq** [comment]
Even if the performance would be reduced by 5% this change would be worth it, IMHO.

**Comment by timotheecour** [comment]
> Even if the performance would be reduced by 5% this change would be worth it, IMHO.

agreed (but to make sure it's 5% and not, say, 40% it requires apples-to-apples comparison and measuring on release/danger builds of nim); I've verified that performance difference is within measurement noise (ie no change), via:
```
nim doc --doccmd:skip --lib:lib -f lib/system.nim
nim rst2html --doccmd:skip -f doc/manual.rst
```
comparing with nim built with -d:release and -d:danger, both before and after this PR. So, all good (pending actually reviewing the code...)

**Comment by Araq** [comment]
I did review it.

File: compiler/docgen.nim [hunk 1]
```diff
@@ -1224,41 +1166,41 @@ proc genSection(d: PDoc, kind: TSymKind, groupedToc = false) =
...
-  if d.section[kind] == nil: return
-  var title = sectionNames[kind].rope
-  d.section[kind] = ropeFormatNamedVars(d.conf, getConfigVar(d.conf, "doc.section"), [
-      "sectionid", "sectionTitle", "sectionTitleID", "content"], [
-      ord(kind).rope, title, rope(ord(kind) + 50), d.section[kind]])
+  if d.section[kind] == "": return
+  var title = sectionNames[kind]
+  d.section[kind] = getConfigVar(d.conf, "doc.section") % [
+      "sectionid", $ord(kind), "sectionTitle", title,
+      "sectionTitleID", $(ord(kind) + 50), "content", d.section[kind]]
```
Comment (timotheecour) [inline compiler/docgen.nim:1173]: pre-existing but (say in future work) we should get rid of this magic 50 repeated in a few places and replace it by a proper const

File: compiler/docgen.nim [hunk 2]
```diff
@@ -1407,14 +1358,16 @@ proc commandJson*(cache: IdentCache, conf: ConfigRef) =
...
   if optStdout in d.conf.globalOptions:
-    writeRope(stdout, content)
+    write(stdout, content)
   else:
     #echo getOutFile(gProjectFull, JsonExt)
     let filename = getOutFile(conf, RelativeFile conf.projectName, JsonExt)
-    if not writeRope(content, filename):
+    try:
+      writeFile(filename, content)
+    except:
```
Comment (timotheecour) [inline compiler/docgen.nim:1370]:
```suggestion
    except IOError:
```

File: compiler/docgen.nim [hunk 3]
```diff
@@ -1427,33 +1380,36 @@ proc commandTags*(cache: IdentCache, conf: ConfigRef) =
...
   if optStdout in d.conf.globalOptions:
-    writeRope(stdout, content)
+    write(stdout, content)
   else:
     #echo getOutFile(gProjectFull, TagsExt)
     let filename = getOutFile(conf, RelativeFile conf.projectName, TagsExt)
-    if not writeRope(content, filename):
+    try:
+      writeFile(filename, content)
+    except:
```
Comment (timotheecour) [inline compiler/docgen.nim:1393]:
```suggestion
    except IOError:
```

File: compiler/docgen.nim [hunk 4]
```diff
@@ -1427,33 +1380,36 @@ proc commandTags*(cache: IdentCache, conf: ConfigRef) =
...
+      "title", "Index",
+      "subtitle", "", "tableofcontents", "", "moduledesc", "",
+      "date", getDateStr(), "time", getClockStr(),
+      "content", content, "author", "", "version", "", "analytics", ""]
   # no analytics because context is not available
 
-  if not writeRope(code, filename):
+  try:
+    writeFile(filename, code)
+  except:
```
Comment (timotheecour) [inline compiler/docgen.nim:1414]:
```suggestion
  except IOError:
```

**Review by timotheecour — APPROVED** [review]
LGTM (remaining points can be addressed later); how about doing the same for cgen and jsgen?

this could be done gradually in several PR's (and making sure performance doesn't drop at least not more than reasonable, since cgen is more critical than nim doc as in this PR)

**Comment by Araq** [comment]
Iirc @arnetheduck did it for the cgen once and got measurably worse performance. What's worse, the cgen should **embrace** the ropes data structure and mutate it into a proper C AST. This can easily be done incrementally, introduce a `kind` enum for the Rope tree and turn into a semi-structured thing that becomes more and more structured in time.

**Comment by timotheecour** [comment]
> Iirc @arnetheduck did it for the cgen once and got measurably worse performance

do you have any links? We should document those experiments.

> What's worse, the cgen should embrace the ropes data structure and mutate it into a proper C AST. 

I agree that we should migrate to AST to AST transformation, so long it keeps handling `.emit`, which are a necessary escape hatch. It should be possible to accommodate `.emit` even with nim AST to C/C++/js AST transformations.

But that doesn't necessarily involve ropes though.

The next step here would be a minimal proof of concept of nim AST => C AST that has some chance of generalizing to js AST and cpp AST (it's not clear that an intermediate representation that can generate c, js, cpp would be practical)

**Comment by arnetheduck** [comment]
> Iirc @arnetheduck did it for the cgen once and got measurably worse performance

not quite, I started doing it but realized that a) refactoring was difficult to get merged into the compiler and b) that a partial move to string will indeed make matters worse - a two-step process would be needed where you first move to `string` _then_ use a better buffer structure for the output - in particular a flat seq instead of ropes - this will be _a lot_ faster and memory efficient than ropes, in general.

A C AST would be easier to post-process massage, indeed - or a MIR-style middle-layer IR that still can be transformed but gets rids of some of the language sugar - such an AST could be used by both C and LLVM backends - either option would be easier to generate sane code from.

**Comment by arnetheduck** [comment]
the best performance could probably be had from SSO for small things and a flat buffer for the output - ropes, but also nim strings in general waste incredible amounts of memory on length, capacity and indirection, most of which is never used - it's generally not reasonable to expect that the same data structure will be efficient for 5-8 character identifiers and 200kb C outputs.

**Comment by Araq** [comment]
> But that doesn't necessarily involve ropes though.

No, but "Rope" is just a name, crucially it's a separate type already and not a string, this property is gold for refactorings, don't remove this just because a string would be faster...

**Pull request merged**